### ISSUE TITLE
[C++] Question : Policy Gradient Loss Function

### ISSUE THREAD

**Issue by Eomtaesan** [Microsoft/CNTK#3379]
1. Variable Definition

reward = CNTK::InputVariable ( { 1 } , false , CNTK::DataType::Float , true , L"reward" );
policy = // output of fully connected layer with 2048 neurons
policyLabel = // external input as label of policy, also has the shape [2048]

2. Loss Functions

[1] // many github examples use this, but does not work
policyLoss = CNTK::CrossEntropyWithSoftmax ( policy , policyLabel );
policyLoss = CNTK::ElementTimes ( policyLoss , reward , L"policyLoss" );

[2] // I makes this function, but I am not sure
policyLoss = CNTK::Log ( CNTK::ReduceSum ( CNTK::ElementTimes ( CNTK::Softmax ( policy ) , policyLabel ) , CNTK::Axis ( 0 ) ) );
policyLoss = CNTK::ElementTimes ( policyLoss , reward , L"policyLoss" );

[3] // policy gradient tutorial of cntk python uses this
policyLoss = CNTK::ReduceMean ( CNTK::Log ( CNTK::SquaredError( CNTK::Softmax ( policy ) , policyLabel ) , CNTK::Axis ( 0 ) ) );
policyLoss = CNTK::ElementTimes ( policyLoss , reward , L"policyLoss" );

what is the correct loss function?

**Comment by tangyuq** [comment]
It depends on your assumptions over the action space. If the policy \pi(a_t | s_t) is assumed to generate a discrete distribution over actions (discrete actions), cross-entropy is the form; if the policy \pi(a_t | s_t) are assumed Gaussion over continuous action space, squared error is the form. 

For policy gradient, the goal is to maximize the expected rewards given a policy J(W) = E_{\tau ~ p(\tau; W)} [ r(\tau) ]. With a few math tricks, the gradient of J(W) can be written as \nabla J(W) = \sum_{t \geq 0} r(\tau) \nabla_{W} log p(a_t | s_t; W).  With one-hot representation of action, the form of log p(a_t | s_t; W)   for softmax discrete action is [1]; the form of log p(a_t | s_t; W)   for a Gaussion over continuous actions is   [2] --- a  squared error. 

There are additional tricks which will further modify the loss. To be precise, it is the gradient of the loss.   Please check paper: https://papers.nips.cc/paper/1713-policy-gradient-methods-for-reinforcement-learning-with-function-approximation.pdf, or refer to any textbooks/lecture notes on deep policy gradient descent.
